integration/cli_test.go

```diff
@@ -112,12 +128,75 @@ func runTestScenariosWithInput(t *testing.T, tests []testCase, input *os.File) {
 // Used for processing test output before comparing to golden files.
 type processOutput func(string) string
 
+// Used for additional logic before executing oauth2l's command.
+type preCommandLogic func(tc *testCase, loopbackAddr string) error
+
+// Used for additional logic after executing oauth2l's command.
+type postCommandLogic func(loopbackAddr string)
+
+// Runs tests where extra logic needs to be added before/after the command execution.
+//
+// input processOutput: processes test output before comparing to golden files
+// input preCmdLogic: additional logic excuted once per test case before command excution. It allows for test argument manipulation.
+// input postCmdLogic: additional logic excuted once per test case after command excution. It allows for cleanup.
+func runTestScenariosWithAdvancedLogic(t *testing.T, tests []testCase, input *os.File, processOutput processOutput,
+	preCmdLogic preCommandLogic, postCmdLogic postCommandLogic) {
+	// Looking for available port.
+	// The port is passed to the the advance logic functions.
+	l, a, err := util.GetListener("http://localhost")
+	if err != nil {
+		t.Fatalf("Error when getting listener: %v", err)
+	}
+	(*l).Close()
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			// Processing logic before exec. Command
+			if preCmdLogic != nil {
+				if err := preCmdLogic(&tc, a); err != nil {
+					t.Fatalf("Unexpected error: %v", err)
+				}
+			}
+
+			cmd := exec.Command(binaryPath, tc.args...)
+			if input != nil {
+				cmd.Stdin = input
+			}
+
+			// exec.Command and gathering output
+			output, err := cmd.CombinedOutput()
+
+			// Processing logic after exec. Command
+			if postCmdLogic != nil {
+				postCmdLogic(a)
+			}
+			if (err != nil) != tc.wantErr {
+				t.Fatalf("%s\nexpected (err != nil) to be %v, but got %v. err: %v", output, tc.wantErr, err != nil, err)
+			}
+			actual := string(output)
+
+			if processOutput != nil {
+				actual = processOutput(actual)
+			}
+
+			golden := newGoldenFile(t, tc.golden)
+
+			if *update {
+				golden.write(actual)
+			}
+			expected := golden.load()
+			if !reflect.DeepEqual(expected, actual) {
+				t.Fatalf("Expected: %v Actual: %v", expected, actual)
+			}
+		})
+	}
+}
+
 // Runs test cases where stdin input is needed and output needs to be processed before comparing to golden files.
 func runTestScenariosWithInputAndProcessedOutput(t *testing.T, tests []testCase, input *os.File, processOutput processOutput) {
```

Comment:
This function should call runTestScenariosWithAdvancedLogic and pass in nil nil as last two parameter, instead of duplicating the code. (See pattern in runTestScenariosWithInput)

Reply:
Hi Andy, I had followed this pattern before. At some point, the two functions looks so different that I needed to diverge. I have restore the previous logic, which is exactly what you are looking for. Thanks for catching this horror :)